contexts/MembershipContext/src/UseCases/HandlePayPalNotification/HandlePayPalMembershipApplicationNotificationUseCase.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\MembershipContext\UseCases\HandlePayPalNotification;
+
+use Psr\Log\LoggerInterface;
+use WMDE\Fundraising\Frontend\Infrastructure\TemplateBasedMailer;
+use WMDE\Fundraising\Frontend\MembershipContext\Authorization\ApplicationAuthorizer;
+use WMDE\Fundraising\Frontend\MembershipContext\Domain\Model\Application;
+use WMDE\Fundraising\Frontend\MembershipContext\Domain\Repositories\ApplicationRepository;
+use WMDE\Fundraising\Frontend\MembershipContext\Domain\Repositories\GetMembershipApplicationException;
+use WMDE\Fundraising\Frontend\MembershipContext\Domain\Repositories\StoreMembershipApplicationException;
+use WMDE\Fundraising\Frontend\PaymentContext\Domain\Model\PayPalData;
+use WMDE\Fundraising\Frontend\PaymentContext\Domain\Model\PayPalPayment;
+
+/**
+ * @license GNU GPL v2+
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ * @author Gabriel Birke < gabriel.birke@wikimedia.de >
+ */
+class HandlePayPalMembershipApplicationNotificationUseCase {
+
+	private $repository;
+	private $authorizationService;
+	private $mailer;
+	private $logger;
+
+	public function __construct( ApplicationRepository $repository, ApplicationAuthorizer $authorizationService,
+								 TemplateBasedMailer $mailer, LoggerInterface $logger ) {
+		$this->repository = $repository;
+		$this->authorizationService = $authorizationService;
+		$this->mailer = $mailer;
+		$this->logger = $logger;
+	}
+
+	public function handleNotification( PayPalNotificationRequest $request ): bool {
+		if ( !$this->requestCanBeHandled( $request ) ) {
+			return false;
+		}
+
+		try {
+			$membershipApplication = $this->repository->getApplicationById( $request->getApplicationId() );
+		} catch ( GetMembershipApplicationException $ex ) {
+			return false;
+		}
+
+		return $this->handleRequestForMembershipApplication( $request, $membershipApplication );
+	}
+
+	private function requestCanBeHandled( PayPalNotificationRequest $request ): bool {
+		if ( $request->isSuccessfulPaymentNotification() ) {
+			return true;
+		}
+
+		if ( $this->transactionIsSubscriptionRelatedButNotAPayment( $request ) ) {
+			return true;
+		}
+
+		$this->logUnhandledStatus( $request );
+		return false;
+	}
+
+	private function handleRequestForMembershipApplication( PayPalNotificationRequest $request,
+															Application $application ): bool {
```

Comment:
I know this is copied 1:1 from the donation use case, but I just realized that some logging (probably on notice level) whenever we return `false` would be really helpful. Each `false` means we did not handle an IPN but have no record of it not being handled.